### PR TITLE
Allow mouse pasting in mouse mode with shift

### LINKED ISSUE #1298: Can’t paste selection with Shift Insert when in mouse mode
It happens since #1218 has been merged.

### PR DESCRIPTION
**Pull request by chrisduerr**
It is now possible to paste in mouse mode again by making use of the
`shift` key while pressing the mouse button reserved for PasteSelection.

This fixes #1298.

### CODE REVIEW

**Comment by chrisduerr** [comment]
This bug raises the question for me if we should really have to check `mods.shift` every time we do something related to mouse-mode. Maybe there would be a better way to do this and we could have a function like `ctx.mouse_mode_enabled(mods)` to have a standard way to get the mouse mode information.

If `mods` is not available `Default::default()` could be passed, I'm pretty sure this would improve the situation by quite a bit and might even handle some cases that are still not behaving properly. Any opinion on that @jwilm? Could probably even implement that in this PR.

**Comment by jwilm** [comment]
The way bindings are setup, it should be possible to handle shift checks without any code changes. The thing I think is missing is a terminal mode flag for mouse mode in binding config. Note that there are methods `Binding::mods_match` and `Binding::mode_matches` to check these conditions. Actions shouldn't actually be dependent on terminal state. The last PR was actually a mistake in that regard.

In summary, we should

1. Remove any mode/mods check from Actions
2. Add a mouse mode option to binding config
3. Add a new default binding for Shift + MiddleMouse when mode contains MouseMode.

**Comment by chrisduerr** [comment]
So what you would say is instead of having this handled in code, we'd allow the user to setup bindings that are for mouse mode only or for mouse mode and "normal" mode?

First of all, right now the modifiers aren't handled at all with mouse events as far as I know, but that's a feature that could be easily added.

In general I'm not a huge fan of this idea though. Because we already have a ton of logic in `src/input.rs` which disables mouse mode checks when shift is pressed, I think it would be super confusing if for some scenarios we do the check ourselves and in other cases expect the user to define custom bindings.

To me the general idea seems like this: If there is ***anything*** that is treated specially when mouse mode is disabled, that behavior should be undone when a certain key (default seems to be shift pretty much everywhere) is pressed. That's where my idea of having a method like `ctx.mouse_mode_enabled(mods)` came from. Not only to use it in the Actions, but to use it everywhere in the `input.rs` so whenever we implement new functionality that only works in mouse mode, the user can disable that by pressing `shift`.

**Comment by jwilm** [comment]
> In general I'm not a huge fan of this idea though. Because we already have a ton of logic in src/input.rs which disables mouse mode checks when shift is pressed.

We could investigate whether it's feasible to turn these all into bindings. The features I see doing this are

* Mouse press to clear selection and run mouse_report
* Mouse release to run mouse_report and copy the selection
* Mouse wheel (no shift) which ultimately is calling scroll_terminal. This method has logic to run one of a few things - mouse_report, faux scroll, or actual scroll
* Mouse moved to update selection if button is pressed and depending on mouse mode

I can imagine how most of these might map to bindings. The benefit here is that it puts users in full control over how the mouse triggers certain actions. For example, we could have an action MouseReport and ClearSelection to trigger on a left MousePress. The mouse wheel is a bit tricky, and multi click + drag is a bit tricky. I believe they could all be expressed as bindings though. This could even be done in a backwards compatible fashion! Since only one binding is ever processed, we can simply append default bindings to the mouse binding list. If the user hasn't configured them, they would still be there. If they have configured the same combinations, their bindings would be preferred.

> idea of having a method like `ctx.mouse_mode_enabled(mods)` came from. Not only to use it in the Actions, but to use it everywhere in the `input.rs` so whenever we implement new functionality that only works in mouse mode, the user can disable that by pressing `shift`.

This is an interesting idea, and I like that it would make all of the mouse actions symmetric. I also like that it would solve the immediate problem without too much effort. My main concern here is that we are further special casing mouse handling and moving away from the central abstraction for input actions.

So, how about this for a plan. Short term, we go with your proposal since it should be easy to implement from where we are today. Long term, I think we should consider making all of these mouse actions into bindings -- even if we don't expose them to config. The main benefit here is that input.rs becomes a state manager which delegates to something that executes actions based on the state rather than having both state management and action processing happening together. WDYT?

**Comment by chrisduerr** [comment]
I feel like this all would just lead to a ton of complexity nobody would ever use, a huge default configuration file and a lot of potential areas for bugs to get introduced.

I feel like the only thing people want is "disable mouse mode when X key is pressed". Instead of "when pressing middle mouse and hjk, then shift, I want mouse mode to be ignored for the next 3.5 key presses". Sure, I agree that whenever possible we should give control to the user to let him decide what he wants, but at some point a complex configuration file is just gonna be a burdon rather than a strength.

I don't think "we are further special casing mouse handling" is the case really, compared to the current version we're just unifying the single special case that has already been everywhere, that shift disables mouse mode. That's pretty much the only special case I'm aware of and to the user of any terminal emulator it probably feels like a special case for a good reason, because it kinda is. It's a bit like "if this single button is pressed, you're good, nothing to worry about, mouse mode is completely disabled". 

I don't think going one way first and then doing the other is beneficial, it just seems like a waste of time in the long-term. Implementing either solution should be fairly simple. I'm totally fine with going for your solution if you decide on that, I just don't want users to share 100 lines of mouse bindings only because alacritty doesn't have sane defaults (which is kinda going on with some key combinations). If we would introduce a new thing in mouse mode for example, every user would have to add two new bindings, one for the normal and one for the special action. If special cases are handled automatically, the user wouldn't have to worry about anything.

**Comment by jwilm** [comment]
>  I just don't want users to share 100 lines of mouse bindings only because alacritty doesn't have sane defaults (which is kinda going on with some key combinations).

This is something that's been on my mind as well. Perhaps we should populate a lot of the standard key bindings outside of config and let the config file simply extend/override that.

> I don't think going one way first and then doing the other is beneficial, it just seems like a waste of time in the long-term

Fair point. Let's go with your solution since it seems far less likely to introduce bugs.

**Comment by chrisduerr** [comment]
I don't want to rush any solution here, but just a few more points that might be relevant:

If we want both block selection and shift escapes as mappings, there might be potential for a conflict if the user configures them both to be the same by accident.

Having mouse movement as config setting seems to be difficult, because `moved + clicked + shift + mouse_mode_on` seems like quite the pill to swallow just to get selection to escape properly.

Generally I'm not even sure if it makes sense to configure the key for escaping the mouse mode at all, it seems that most terminal emulators have decided on `shift` for that and I doubt anyone would want anything else. And even if you use something else, there's potential for conflicts with applications, some applications use ctrl+click for example, so it would not be possible to use that mapping without possibly breaking applications.

I have not much expertise in this area, so I wouldn't value my guesses too much, but I'd say the trade-off here is either having a powerful and complex system with potential breakage VS a system that hopefully "just works" and is good enough for everyone. I think with the high degree of standardization around stuff like `shift` to escape mouse modes, we have a good chance that going for the simple and "dumb" solution might do the job, instead of going for what's clever that might eventually just lead to difficulties for maintainers and users.

**Comment by chrisduerr** [comment]
The changes I've introduced are a bit more than I would have originally expected, but I uncovered a few bugs/inconsistencies especially with faux scrolling. So here's the commit message as a summary/explanation of my actions:

```
This change aims to make it simpler to check if the mouse mode is
enabled by introducing a method that automatically checks if shift is
pressed. There were a few bugs I noticed in the current system which
have also been fixed.

When reporting all mouse events and shift is held, mouse events were
still sent, this is inconsistent with xterm, urxvt and termite. This has
been changed to suppress these while the shift key is pressed.

Faux scrolling had some inconsistencies and wasn't really behaving 100%
like the VTE terminal I've used for reference for the original
implementation (VTE). So this PR also fixes some issues that were
present with faux scrolling.

These changes have been made:
    BEFORE:
        vim scroll -> Faux Scrolling
        vim shift + scroll -> Faux Scrolling
        vim :set mouse=a scroll -> Application Scrolling
        vim :set mouse=a shift + scroll -> Page Up/Down
        man scroll -> Faux Scrolling
        man shift + scroll -> Faux Scrolling

    AFTER:
        vim scroll -> Faux Scrolling
        vim shift + scroll -> Page Up/Down
        vim :set mouse=a scroll -> Application Scrolling
        vim :set mouse=a shift + scroll -> Page Up/Down
        man scroll -> Faux Scrolling
        man shift + scroll -> Page Up/Down

    VTE (Termite):
        vim scroll -> Faux Scrolling
        vim shift + scroll -> Page Up/Down
        vim :set mouse=a scroll -> Application Scrolling
        vim :set mouse=a shift + scroll -> Page Up/Down
        man scroll -> Faux Scrolling
        man shift + scroll -> NOOP

`vim` and `man` have been used for testing this, the significance
between their difference is that `man` uses the APP_CURSOR mode.

The main difference between the previous behavior is that SHIFT + FAUX
SCROLLING should now always result in page up/down.

The reason for having a different implementation than my reference
termite is that I do not think their approach makes much sense, so
instead of sending page up/down X times, it's just sent once. And it's
sent in all applications.
```

**Comment by sazarkin** [comment]
hi, can you rebase this on master?

**Comment by chrisduerr** [comment]
@sazarkin Since this introduces some significant changes to the mouse handling code, jwilm will need to review this first. I will rebase this on top of master when it is approved so it can be merged without any issues.

**Comment by l29ah** [comment]
bump?

**Comment by chrisduerr** [comment]
@l29ah This still needs some changes before getting merged and I'm currently focused on getting more recent PRs out of the way. Please rest assured that I do not forget my own PRs.

Unfortunately asking for updates usually just delays things more since I need to respond to github notifications rather than working on the issue itself.

**Comment by chrisduerr** [comment]
I've reworked this PR completely. I've decided to remove the 'abstraction' for checking if the mouse mode is active since it mostly just introduced additional noise anyways.

This should now just simply fix the issue, rather than introducing some big changes. However, mouse modifiers are always matched in a relaxed way (similar to the URL checking), because additional modifiers are commonly used with the mouse.
